File: ruby/lib/woothee/os.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/util'

module Woothee::OS
  extend Woothee::Util

  def self.challenge_windows(ua, result)
    return false if ua.index("Windows").nil?

    data = Woothee::DataSet.get('Win')
    unless ua =~ /Windows ([ .a-zA-Z0-9]+)[;\\)]/o
      # Windows, but version unknown
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_NAME])
      return true
    end

    version = $1
    data = case
           when version == 'NT 6.3' then Woothee::DataSet.get('Win8.1')
           when version == 'NT 6.2' then Woothee::DataSet.get('Win8')
           when version == 'NT 6.1' then Woothee::DataSet.get('Win7')
           when version == 'NT 6.0' then Woothee::DataSet.get('WinVista')
           when version == 'NT 5.1' then Woothee::DataSet.get('WinXP')
           when version =~ /^Phone OS/o then Woothee::DataSet.get('WinPhone')
           when version == 'NT 5.0' then Woothee::DataSet.get('Win2000')
           when version == 'NT 4.0' then Woothee::DataSet.get('WinNT4')
           when version == '98' then Woothee::DataSet.get('Win98') # wow, WinMe is shown as 'Windows 98; Win9x 4.90', fxxxk
           when version == '95' then Woothee::DataSet.get('Win95')
           when version == 'CE' then Woothee::DataSet.get('WinCE')
           else
             data # windows unknown version
           end
    update_category(result, data[Woothee::KEY_CATEGORY])
    update_os(result, data[Woothee::KEY_NAME])
    true
  end

  def self.challenge_osx(ua, result)
    return false if ua.index('Mac OS X').nil?

    data = Woothee::DataSet.get('OSX')
    if ua.index('like Mac OS X')
      data = case
             when ua.index('iPhone;') then Woothee::DataSet.get('iPhone')
             when ua.index('iPad;') then Woothee::DataSet.get('iPad')
             when ua.index('iPod;') then Woothee::DataSet.get('iPod')
             else data
             end
    end
    update_category(result, data[Woothee::KEY_CATEGORY])
    update_os(result, data[Woothee::KEY_NAME])
    true
  end

  def self.challenge_linux(ua, result)
    return false if ua.index('Linux').nil?

    data = if ua.index('Android')
             Woothee::DataSet.get('Android')
           else
             Woothee::DataSet.get('Linux')
           end
    update_category(result, data[Woothee::KEY_CATEGORY])
    update_os(result, data[Woothee::KEY_NAME])
    true
  end

  def self.challenge_smartphone(ua, result)
    data = case
           when ua.index('iPhone') then Woothee::DataSet.get('iPhone')
           when ua.index('iPad') then Woothee::DataSet.get('iPad')
           when ua.index('iPod') then Woothee::DataSet.get('iPod')
           when ua.index('Android') then Woothee::DataSet.get('Android')
           when ua.index('CFNetwork') then Woothee::DataSet.get('iOS')
           when ua.index('BlackBerry') then Woothee::DataSet.get('BlackBerry')
           else nil
           end

    if result[Woothee::KEY_NAME] && result[Woothee::KEY_NAME] == Woothee::DataSet.get('Firefox')[Woothee::KEY_NAME]
        # Firefox OS specific pattern
        # http://lawrencemandel.com/2012/07/27/decision-made-firefox-os-user-agent-string/
      if ua =~ /^Mozilla\/[.0-9]+ \(Mobile;(.*;)? rv:[.0-9]+\) Gecko\/[.0-9]+ Firefox\/[.0-9]+$/
        data = Woothee::DataSet.get('FirefoxOS')
      end
    end

    return false unless data

    update_category(result, data[Woothee::KEY_CATEGORY])
    update_os(result, data[Woothee::KEY_NAME])
    true
  end

  def self.challenge_mobilephone(ua, result)
    if ua.index('KDDI-')
      if ua =~ /KDDI-([^- \/;()"']+)/o
        term = $1
        data = Woothee::DataSet.get('au')
        update_category(result, data[Woothee::KEY_CATEGORY])
        update_os(result, data[Woothee::KEY_OS])
        update_version(result, term)
        return true
      end
    end
    if ua.index('WILLCOM') or ua.index('DDIPOCKET')
      if ua =~ /(?:WILLCOM|DDIPOCKET);[^\/]+\/([^ \/;()]+)/o
        term = $1
        data = Woothee::DataSet.get('willcom')
        update_category(result, data[Woothee::KEY_CATEGORY])
        update_os(result, data[Woothee::KEY_OS])
        update_version(result, term)
        return true
      end
    end
    if ua.index('SymbianOS')
      data = Woothee::DataSet.get('SymbianOS')
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_OS])
      return true
    end
    if ua.index('Google Wireless Transcoder')
      update_map(result, Woothee::DataSet.get('MobileTranscoder'))
      update_version(result, 'Google')
      return true
    end
    if ua.index('Naver Transcoder')
      update_map(result, Woothee::DataSet.get('MobileTranscoder'))
      update_version(result, 'Naver')
      return true
    end

    false
  end

  def self.challenge_appliance(ua, result)
    if ua.index('Nintendo DSi;')
      data = Woothee::DataSet.get('NintendoDSi')
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_OS])
      return true
    end
    if ua.index('Nintendo Wii;')
      data = Woothee::DataSet.get('NintendoWii')
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_OS])
      return true
    end

    false
  end

  def self.challenge_misc(ua, result)
    data = case
           when ua.index('(Win98;')
             Woothee::DataSet.get('Win98')
           when ua.index('Macintosh; U; PPC;')
             Woothee::DataSet.get('MacOS')
           when ua.index('Mac_PowerPC')
             Woothee::DataSet.get('MacOS')
           when ua.index('X11; FreeBSD ')
             Woothee::DataSet.get('BSD')
           else
             nil
           end
    if data
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_NAME])
      return true
    end

    false
  end
end

File: python/lib/woothee/os.py
# -*- coding: utf-8 -*-

import re
from . import dataset
from . import util

def challenge_windows(ua, result):
  if 'Windows' not in ua:
    return False
  data = dataset.get('Win')
  obj = re.search('Windows ([ .a-zA-Z0-9]+)[;\\)]', ua)
  if not obj:
    # Windows, but version unknown
    util.update_category(result, data[dataset.KEY_CATEGORY])
    util.update_os(result, data[dataset.KEY_NAME])
    return True
  
  version = obj.group(1)
  if version == 'NT 6.3':
    data = dataset.get('Win8.1')
  elif version == 'NT 6.2':
    data = dataset.get('Win8')
  elif version == 'NT 6.1':
    data = dataset.get('Win7')
  elif version == 'NT 6.0':
    data = dataset.get('WinVista')
  elif version == 'NT 5.1':
    data = dataset.get('WinXP')
  elif re.search('^Phone OS', version):
    data = dataset.get('WinPhone')
  elif version == 'NT 5.0':
    data = dataset.get('Win2000')
  elif version == 'NT 4.0':
    data = dataset.get('WinNT4')
  elif version == '98':
    data = dataset.get('Win98') # wow, WinMe is shown as 'Windows 98; Win9x 4.90', fxxxk
  elif version == '95':
    data = dataset.get('Win95')
  elif version == 'CE':
    data = dataset.get('WinCE')
  
  util.update_category(result, data[dataset.KEY_CATEGORY])
  util.update_os(result, data[dataset.KEY_NAME])
  return True

def challenge_osx(ua, result):
  if 'Mac OS X' not in ua:
    return False
  data = dataset.get('OSX')
  if 'like Mac OS X' in ua:
    if 'iPhone;' in ua:
      data = dataset.get('iPhone')
    elif 'iPad;' in ua:
      data = dataset.get('iPad')
    elif 'iPod;' in ua:
      data = dataset.get('iPod')
  util.update_category(result, data[dataset.KEY_CATEGORY])
  util.update_os(result, data[dataset.KEY_NAME])
  return True

def challenge_linux(ua, result):
  if 'Linux' not in ua:
    return False
  if 'Android' in ua:
    data = dataset.get('Android')
  else:
    data = dataset.get('Linux')
  util.update_category(result, data[dataset.KEY_CATEGORY])
  util.update_os(result, data[dataset.KEY_NAME])
  return True

def challenge_smartphone(ua, result):
  data = None
  if 'iPhone' in ua:
    data = dataset.get('iPhone')
  elif 'iPad' in ua:
    data = dataset.get('iPad')
  elif 'iPod' in ua:
    data = dataset.get('iPod')
  elif 'Android' in ua:
    data = dataset.get('Android')
  elif 'CFNetwork' in ua:
    data = dataset.get('iOS')
  elif 'BlackBerry' in ua:
    data = dataset.get('BlackBerry')

  if dataset.KEY_NAME in result and result[dataset.KEY_NAME] == dataset.get('Firefox')[dataset.KEY_NAME]:
    # Firefox OS specific pattern
    # http://lawrencemandel.com/2012/07/27/decision-made-firefox-os-user-agent-string/
    if re.search('^Mozilla/[.0-9]+ \(Mobile;(.*;)? rv:[.0-9]+\) Gecko/[.0-9]+ Firefox/[.0-9]+$', ua):
      data = dataset.get('FirefoxOS')

  if not data:
    return False

  util.update_category(result, data[dataset.KEY_CATEGORY])
  util.update_os(result, data[dataset.KEY_NAME])
  return True

def challenge_mobilephone(ua, result):
  if 'KDDI-' in ua:
    obj = re.search('KDDI-([^- /;()"\']+)', ua)
    if obj:
      term = obj.group(1)
      data = dataset.get('au')
      util.update_category(result, data[dataset.KEY_CATEGORY])
      util.update_os(result, data[dataset.KEY_OS])
      util.update_version(result, term)
      return True
  if 'WILLCOM' in ua or 'DDIPOCKET' in ua:
    obj = re.search('(?:WILLCOM|DDIPOCKET);[^/]+/([^ /;()]+)', ua)
    if obj:
      term = obj.group(1)
      data = dataset.get('willcom')
      util.update_category(result, data[dataset.KEY_CATEGORY])
      util.update_os(result, data[dataset.KEY_OS])
      util.update_version(result, term)
      return True
  if 'SymbianOS' in ua:
    data = dataset.get('SymbianOS')
    util.update_category(result, data[dataset.KEY_CATEGORY])
    util.update_os(result, data[dataset.KEY_OS])
    return True
  if 'Google Wireless Transcoder' in ua:
    util.update_map(result, dataset.get('MobileTranscoder'))
    util.update_version(result, 'Google')
    return True
  if 'Naver Transcoder' in ua:
    util.update_map(result, dataset.get('MobileTranscoder'))
    util.update_version(result, 'Naver')
    return True
  return False

def challenge_appliance(ua, result):
  if 'Nintendo DSi;' in ua:
    data = dataset.get('NintendoDSi')
    util.update_category(result, data[dataset.KEY_CATEGORY])
    util.update_os(result, data[dataset.KEY_OS])
    return True
  if 'Nintendo Wii;' in ua:
    data = dataset.get('NintendoWii')
    util.update_category(result, data[dataset.KEY_CATEGORY])
    util.update_os(result, data[dataset.KEY_OS])
    return True
  return False

def challenge_misc(ua, result):
  if '(Win98;' in ua:
    data = dataset.get('Win98')
  elif 'Macintosh; U; PPC;' in ua:
    data = dataset.get('MacOS')
  elif 'Mac_PowerPC' in ua:
    data = dataset.get('MacOS')
  elif 'X11; FreeBSD ' in ua:
    data = dataset.get('BSD')
  else:
    return False
  util.update_category(result, data[dataset.KEY_CATEGORY])
  util.update_os(result, data[dataset.KEY_NAME])
  return True
